OKTA-456536 Secure API connections between orgs with OAuth 2.0 (#2992)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/secure-oauth-between-orgs/main/index.md

* Configure Org2Org app on each spoke Okta org.
* Configure and activate Org2Org provisioning on spoke Okta orgs.
* Rotate keys for the OAuth 2.0 connection.

**What you need**

* Multiple Okta orgs for your multi-tenant solution

---

## About using OAuth 2.0 for secure API connections in multi-tenant solutions

To secure API connections between orgs for a [multi-tenant solution](/docs/concepts/multi-tenancy/) that uses the hub and spoke model, Okta provides the OAuth 2.0 client credential flow using the Okta Org2Org integration as an OAuth 2.0 client. This enables Okta admins to push and match both users and groups from one Okta org to another. Previously, this integration only supported token-based access to the Okta API. Now the Org2Org integration can be configured to access the Okta API as an OAuth 2.0 client. This increases security by limiting the scope of access and providing a better mechanism to rotate credentials.

![Displays the service apps required in the hub org and the Org2Org apps required in the spoke org.](/img/multi-tenancy/multi-org-provisioning.png "OAuth 2.0 connection configuration for hub and spoke org model")

> In this OAuth 2.0 flow, the hub org acts as the resource server as well as the [authorization server](/docs/concepts/auth-servers/).  Each spoke org is represented by a [service app in the hub org for the OAuth 2.0 client credentials flow](/docs/guides/implement-oauth-for-okta-serviceapp/main/). At the spoke org level, you need to create an Org2Org app to represent the OAuth 2.0 client. The Okta Org2Org app automatically generates a JSON Web Key Set (JWKS)  used by the client to authenticate with the authorization server. For each Org2Org app, you need to extract the JWKS public key for each corresponding service app in the hub org. To limit access to the OAuth 2.0 clients, you need to grant the hub org service app with allowed [OAuth 2.0 scopes](/docs/guides/implement-oauth-for-okta/main/#scopes-and-supported-endpoints).

Author: good 👀 ! 👍

After the Org2Org OAuth 2.0 connection is configured, you can start to test your connection and push user data from the spoke orgs to the hub org. You can also create a [rotate keys](#key-rotation) schedule to update your credentials on a regular basis.

> > **Note**: For a hub and spoke org model provisioning using API tokens instead of OAuth 2.0, see [Integrate Okta Org2Org with Okta ](https://help.okta.com/okta_help.htm?id=ext-org2org-intg).

Reviewer: Maybe:

"For provisioning in a hub and spoke org model that uses API tokens instead of OAuth 2.0...

Author: Sounds good. Thanks!  👍

## Hub and spoke connection configuration with OAuth 2.0

> You can use OAuth 2.0 to push user and group information from a spoke org to a centralized hub org by performing the following configuration:

Reviewer: is 'centralized' redundant?

Author: Possibly redundant. But I wanted to stress the central org here.

1. In each spoke org, [add an instance of the Org2Org app integration](#add-an-org2org-app-integration-in-a-spoke-org).

> 2. In the hub org, [create an OAuth 2.0 service app](#create-an-oauth-2-0-service-app-in-the-hub-org) with `private_key_jwt` authentication method and JWKS public key of each Org2Org app in the spoke orgs. Grant [allowed scopes](#request-examples) to the hub org service app (the OAuth 2.0 client).

Reviewer: maybe "...with the 'private... ...and the JWKS public..." (add two 'the's?)

3. [Set and activate provisioning in the Org2Org apps](#enable-provisioning-in-the-org2org-app) from the Okta API.
4. [Assign users and groups to be pushed to the hub org](#assign-users-and-groups-in-the-org2org-app).

### Add an Org2Org app integration in a spoke org

In the spoke org, where you want to push users and groups to the central hub org, you need to add an instance of the Org2Org app integration by using the [Okta Apps API](/docs/reference/api/apps/#add-basic-authentication-application). This generates an Org2Org app integration instance and the key certificates required to connect to the hub org.

As an Okta admin, make a `POST /api/v1/apps` request to the spoke org with [Okta Org2Org parameters](/docs/reference/api/apps/#add-okta-org2org-application):

| Parameter |  Description/Value   |
| --------- |  ------------- |
| `name`  |  `okta_org2org` |
| `label`  |  Specify a label for this Org2Org app integration. |

> | `baseUrl`  |  Specify your hub org base URL. |

Reviewer: Maybe:
Specify the base URL of your hub org.
or
The base URL of your hub org
?

-d '{
  "name": "okta_org2org",
  "label": "My spoke Org2Org App",
  "signOnMode": "SAML_2_0",
  "settings": {
    "app": {
        "baseUrl": "https://${yourHubOktaDomain}"
    }
  }
}' "https://${yourSpokeOktaDomain}/api/v1/apps"
```

From the response of your POST request, use the `id` property of the Org2Org app instance to [retrieve the key credentials](/docs/reference/api/apps/#list-key-credentials-for-application) generated for the app with the `GET /api/v1/apps/${id}/credentials/keys` request.

##### Request example

```bash
curl -v -X GET \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
"https://${yourOktaDomain}/api/v1/apps/${yourOrg2OrgAppId}/credentials/keys"
```

##### Response example

```json
[
  {
    "created": "2015-12-10T18:55:35.000Z",
    "expiresAt": "2045-01-23T02:15:23.000Z",
    "x5c": [
      "MIIDqDCCApCgAwIBAgIGAUsUko … llBooPPS3wYlNA=="
    ],
    "e": "AQAB",
    "n": "htbi5H5MN_oYaKcZ … hVagjTzo",
    "kty": "RSA",
    "use": "sig",
    "x5t#S256": "7CCyXWwKzH4P6PoBP91B1S_iIZVzuGffVnUXu-BTYQQ"
  }
]
```

> **Note**: The keys are truncated for brevity.

> Save the returned key credentials for use in the hub org service app configuration.

Reviewer: Maybe:
Save the returned key credentials for use in the service app configuration for the hub org.
?

    "profile": {
        "authScheme": "OAUTH2",
        "clientId": "${yourHubOrgServiceAppId}"
    }
}' "https://${yourSpokeOrgDomain}/api/v1/apps/${yourOrg2OrgAppId}/connections/default?activate=TRUE"
```

> **Note**: After provisioning is enabled, if you want to edit Org2Org attribute mappings, you can use the [App API features operation](/docs/reference/api/apps/#list-features-for-application) and the [Mappings API](/docs/reference/api/mappings/), or alternatively, you can go to the Okta Admin Console and edit the **Okta Org2Org Attribute Mappings** section of the Org2Org app **Provisioning To App** settings.

### Assign users and groups in the Org2Org app

In each spoke org, assign the users and groups to the Org2Org app integration by using the [Okta Apps API application operations](/docs/reference/api/apps/#application-user-operations):

* [`POST /api/v1/apps/${yourOrg2OrgAppId}/users` (assign user to an application for SSO and provisioning)](/docs/reference/api/apps/#assign-user-to-application-for-sso-and-provisioning)
* [`POST /api/v1/apps/${yourOrg2OrgAppId}/groups/${groupId}` (assign group to an application)](/docs/reference/api/apps/#assign-group-to-application)

> Alternatively, you can assign users and groups for provisioning using the Okta Admin Console. See [ Assign an app integration to a user](https://help.okta.com/oie/en-us/Content/Topics/Provisioning/lcm/lcm-assign-app-user.htm) and [Assign an app integration to a group](https://help.okta.com/oie/en-us/Content/Topics/Provisioning/lcm/lcm-assign-app-groups.htm).

Reviewer: Not sure if it matters, but might want to remove the space after the [ opening bracket.

---
title: Secure API connections between orgs with OAuth 2.0
excerpt: Learn how to synchronize users and groups between orgs with OAuth 2.0 in a multi-tenancy solution.
layout: Guides
---

<ApiLifecycle access="ea" />

This guide explains how to securely set up Okta hub and spoke orgs to synchronize users and groups by using OAuth 2.0 in a [multi-tenant solution](/docs/concepts/multi-tenancy/). The connection between Okta orgs can be set up for API access to specific scoped data using the OAuth 2.0 client credential flow for the [Org2Org integration](https://help.okta.com/okta_help.htm?type=oie&id=ext-org2org-intg) app.

> **Note**: Currently, only the Okta API can be used to enable OAuth 2.0-based provisioning.
---

**Learning outcomes**

* Set up a multi-tenant hub and spoke Okta org provisioning configuration to use the OAuth 2.0 client credential grant flow.

> * Configure hub Okta org with service apps for each spoke Okta org.

Reviewer 2: ```suggestion
* Configure the hub Okta org with service apps for each spoke Okta org.
```

> * Configure Org2Org app on each spoke Okta org.

Reviewer 2: ```suggestion
* Configure an Org2Org app on each spoke Okta org.
```

> **Note**: For provisioning in a hub and spoke org model that uses API tokens instead of OAuth 2.0, see [Integrate Okta Org2Org with Okta ](https://help.okta.com/okta_help.htm?type=oie&id=ext-org2org-intg).

## Hub and spoke connection configuration with OAuth 2.0

You can use OAuth 2.0 to push user and group information from a spoke org to a centralized hub org by performing the following configuration:

1. In each spoke org, [add an instance of the Org2Org app integration](#add-an-org2org-app-integration-in-a-spoke-org).
2. In the hub org, [create an OAuth 2.0 service app](#create-an-oauth-2-0-service-app-in-the-hub-org) with the `private_key_jwt` authentication method and the JWKS public key of each Org2Org app in the spoke orgs. Grant [allowed scopes](#request-examples) to the hub org service app (the OAuth 2.0 client).
3. [Set and activate provisioning in the Org2Org apps](#enable-provisioning-in-the-org2org-app) from the Okta API.
4. [Assign users and groups to be pushed to the hub org](#assign-users-and-groups-in-the-org2org-app).

### Add an Org2Org app integration in a spoke org

> In the spoke org, where you want to push users and groups to the central hub org, you need to add an instance of the Org2Org app integration by using the [Okta Apps API](/docs/reference/api/apps/#add-basic-authentication-application). This generates an Org2Org app integration instance and the key certificates required to connect to the hub org.

Author: Thanks! I'll update the link to [Okta Apps API](/docs/reference/api/apps/#add-okta-org2org-application)

As an Okta admin, make a `POST /api/v1/apps` request to the spoke org with [Okta Org2Org parameters](/docs/reference/api/apps/#add-okta-org2org-application):

| Parameter |  Description/Value   |
| --------- |  ------------- |
| `name`  |  `okta_org2org` |

> | `label`  |  Specify a label for this Org2Org app integration. |

Reviewer 2: ```suggestion
| `label`  |  Specify a label for this Org2Org app integration |
```

> | `baseUrl`  |  Specify the base URL of your hub org. |

Reviewer 2: ```suggestion
| `baseUrl`  |  Specify the base URL of your hub org |
```

    }
  }
}' "https://${yourSpokeOktaDomain}/api/v1/apps"
```

From the response of your POST request, use the `id` property of the Org2Org app instance to [retrieve the key credentials](/docs/reference/api/apps/#list-key-credentials-for-application) generated for the app with the `GET /api/v1/apps/${id}/credentials/keys` request.

##### Request example

```bash
curl -v -X GET \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${spokeApiToken}" \
"https://${yourOktaDomain}/api/v1/apps/${yourOrg2OrgAppId}/credentials/keys"
```

##### Response example

```json
[
    {
        "kty": "RSA",
        "created": "2022-01-20T19:50:14.000Z",
        "lastUpdated": "2022-01-20T19:50:14.000Z",
        "expiresAt": "2024-01-20T19:50:13.000Z",
        "kid": "sf-jWwRKMUU55 ... GucHLxIh_-fYLAofB8",
        "use": "sig",
        "x5c": [
            "MIIDqDCCApCgAwIBAgIGAX55CiDiMA0GCSqG ... c5Iuo9j3wpemDSgGapXQ=="
        ],
        "x5t#S256": "v-v2V8soFmXuhC ... nrJ4ho-N3P8aASFc",
        "e": "AQAB",
        "n": "gIxwqCNkdAb1ioyNBY2boqUCrMj_NSFJAl ... 7dZFiAYF7p_k3XMXOh-hsL_D8FDQ"
    }
]
```

> **Note**: The keys are truncated for brevity.

> You need to save the JWKS properties `kty`, `e`, `n`, `use`, and `kid` for use in the corresponding hub org service app.

Author: Ok, I will remove this sentence and copy the whole response to the service app.

> For each spoke org that you have in your multi-tenant solution, you must create an Org2Org app instance and generate the key credentials that need to be configured in the hub org’s corresponding service app.

Reviewer 3: can we update it to `you must create an Org2Org app instance which comes with auto generated key credentials that...`

Author: I decided to remove this line. It seems repetitive?

### Create an OAuth 2.0 service app in the hub org

 In the hub org, create an OAuth 2.0 service app for each spoke org by using the [Dynamic Client Registration API](/docs/reference/api/oauth-clients/). As an Okta admin, make a `POST /oauth2/v1/clients` request to the hub org with the following required parameters:

| Parameter |  Description/Value   |
| --------- |  ------------- |

> | `client_name`  |  Specify a label for this service app to represent the spoke org OAuth 2.0 client. |

Reviewer 2: ```suggestion
| `client_name`  |  Specify a label for this service app to represent the spoke org OAuth 2.0 client |
```

| `grant_types`  |  `client_credentials` |

> | `jwks`  |  Specify the JSON Web Key Set from the corresponding spoke org’s Org2Org app integration. |

Reviewer 2: ```suggestion
| `jwks`  |  Specify the JSON Web Key Set from the corresponding spoke org’s Org2Org app integration |
```

  -d ' {
    "scopeId": "okta.users.manage",
    "issuer": "https://'${yourHubOrgDomain}'"
}' "https://${yourHubOrgDomain}/api/v1/apps/${yourServiceAppId}/grants"
```

```bash
curl -X POST \
  -H 'Accept: application/json' \
  -H "Authorization: SSWS ${hubApiToken}" \
  -H 'Content-Type: application/json' \
  -d ' {
    "scopeId": "okta.groups.manage",
    "issuer": "https://'${yourHubOrgDomain}'"
}' "https://${yourHubOrgDomain}/api/v1/apps/${yourServiceAppId}/grants"
```

### Enable provisioning in the Org2Org app

In each spoke org, set and activate provisioning for the Org2Org app integration by using the [Okta Apps API](/docs/reference/api/apps/#set-default-provisioning-connection-for-application).

> **Note**: Currently, only the Okta API can be used to enable OAuth 2.0-based provisioning.

As an Okta admin, make a [`POST /api/v1/apps/${Org2OrgAppId}/connections/default?activate=TRUE`](/docs/reference/api/apps/#set-default-provisioning-connection-for-application) request to set provisioning for the spoke org using the following profile parameters:

| Parameter |  Description/Value   |
| --------- |  ------------- |
| `authScheme`  |  `OAUTH2` |

> | `clientId`  |  Specify the corresponding service app client ID in your hub org. |

Reviewer 2: ```suggestion
| `clientId`  |  Specify the corresponding service app client ID in your hub org |
```

##### Request example

```bash
curl -X POST \
  -H 'Accept: application/json' \
  -H "Authorization: SSWS ${spokeApiToken}" \
  -H 'Content-Type: application/json' \
  -d ' {
    "profile": {
        "authScheme": "OAUTH2",
        "clientId": "'${yourHubOrgServiceAppId}'"
    }
}' "https://${yourSpokeOrgDomain}/api/v1/apps/${yourOrg2OrgAppId}/connections/default?activate=TRUE"
```

> > **Note**: After provisioning is enabled, if you want to edit Org2Org attribute mappings, you can use the [App API features operation](/docs/reference/api/apps/#list-features-for-application) and the [Mappings API](/docs/reference/api/mappings/), or alternatively, you can go to the Okta Admin Console and edit the **Okta Org2Org Attribute Mappings** section of the Org2Org app **Provisioning To App** settings.

Reviewer 2: Breaking into two sentences.

```suggestion
> **Note**: After you enable provisioning, if you want to edit Org2Org attribute mappings, you can use the [App API features operation](/docs/reference/api/apps/#list-features-for-application) and the [Mappings API](/docs/reference/api/mappings/). Alternatively, you can go to the Okta Admin Console and edit the **Okta Org2Org Attribute Mappings** section of the Org2Org app **Provisioning To App** settings.
```

}' "https://${yourSpokeOrgDomain}/api/v1/apps/${yourOrg2OrgAppId}"
```

> **Note**: You must specify `name` and `label` body parameters when you update an Org2Org app.

Then go to your hub org and register the new key to your corresponding service app by using the [Dynamic Client Registration API](/docs/reference/api/oauth-clients/).

##### Request example

```bash
curl -X PUT \
  -H 'Accept: application/json' \
  -H "Authorization: SSWS ${hubApiToken}" \
  -H 'Content-Type: application/json' \
  -d ' {
    "client_id": "'${yourServiceAppId}'",
    "client_name": "'${hubServiceClientLabel}'",
    "response_types": [
      "token"
    ],
    "grant_types": [
      "client_credentials"
    ],
    "token_endpoint_auth_method": "private_key_jwt",
    "application_type": "service",
    "jwks": {
                "keys": [
                  {
                    "kty": "'${kty}'",
                    "kid": "'${kid}'",
                    "e": "'${e}'",
                    "alg": "RS256",
                    "use": "'${use}'",
                    "n": "'${n}'"
                  }
                ]
            }
 }' "https://${yourHubOrgDomain}/oauth2/v1/clients/${yourServiceAppId}"
```

> > **Note**: When you update the keys in the service app `jwks` property, all the old keys are overwritten. You must specify all settings when you update a client app. Partial updates aren't supported. If any settings are missing when you update a client app, the update fails.

Reviewer 3: For Key Rotation, if they follow these steps
1. generate new key
2. update Org2Org's kid
3. register new key with service app

there will be some downtime between step 2 and 3
can we swap step 2 and 3:
1. generate new key
2. register new key with service app
3. update Org2Org's kid 

and add a note: If you want to achieve zero downtime for key rotation, you can update service app with both old and new key in step2(`jwks.keys` is an array, different keys are identified by `kid`), then: 
4. (Optional) update service app to remove the old key

Author: I added these notes in the "Key Rotation" section. Thanks for the feedback!

* [`POST /api/v1/apps/${yourOrg2OrgAppId}/groups/${groupId}` (assign group to an application)](/docs/reference/api/apps/#assign-group-to-application)

Alternatively, you can assign users and groups for provisioning using the Okta Admin Console. See [Assign an app integration to a user](https://help.okta.com/okta_help.htm?type=oie&id=ext-lcm-assign-app-user) and [Assign an app integration to a group](https://help.okta.com/okta_help.htm?type=oie&id=ext-lcm-assign-app-groups).

After you've assigned your users or groups in the spoke org, validate that the same users or groups appear in your hub org.

## Key rotation

An advantage to using the OAuth 2.0 connection is that you can [rotate keys](/docs/concepts/key-rotation) to adhere to cryptographic best practices. You can rotate keys for a specific OAuth 2.0 connection by following these API steps:

1. [Generate a new key for the Org2Org app](#generate-a-new-key-for-the-org2org-app)
2. [Register the new key with the corresponding service app](#register-the-new-org2org-app-key-with-the-corresponding-service-app)
3. [Update the current credentials for the Org2Org app](#update-the-current-credentials-for-the-org2org-app)

> > **Note**: If you want to minimize downtime during key rotation, you can update the service app (step two) with both the old and new keys, since `jwks.keys` is an array that can handle different `kid` identifiers. You can remove the old key after you've verified that provisioning works with the new key.

Reviewer 2: ```suggestion
> **Note**: If you want to minimize downtime during key rotation, you can update the service app (step two) with both the old and new keys, since `jwks.keys` is an array that can handle different `kid` identifiers. You can remove the old key after you verify that provisioning works with the new key.
```

### Generate a new key for the Org2Org app

From your spoke org, make a request to [generate a new application key credential](/docs/reference/api/apps/#generate-new-application-key-credential) as an Okta admin user.

##### Request example

```bash
curl -X POST \
  -H 'Accept: application/json' \
  -H "Authorization: SSWS ${spokeApiToken}" \
  -H 'Content-Type: application/json' \
  "https://${yourSpokeOrgDomain}/api/v1/apps/${yourOrg2OrgAppId}/credentials/keys/generate?validityYears=${validYears}"
```

##### Response example

```json
{
    "kty": "RSA",
    "created": "2022-01-20T19:50:14.000Z",
    "lastUpdated": "2022-01-20T19:50:14.000Z",
    "expiresAt": "2024-01-20T19:50:13.000Z",
    "kid": "sf-jWwRKMUU5588aokhj-xu_mGucHLxIh_-fYLAofB8",
    "use": "sig",

>     "x5c": [

Reviewer 2: nit - the formatting of the x5c section should be indented a bit.

Author: Do you want the whole "x5c" line indented? or just the line below?
Btw, I copied this JSON from Postman response and I just shortened the long strings.

Reviewer 2: It's fine - or maybe move the bracket up to the same line as the key? But, it's not incorrect - just that closing bracket threw me. I'm good now. I've had coffee.
